### PR TITLE
gzip: use Java gzip libraries for buffered decoding of gzip files

### PR DESCRIPTION
**Pull request by yaauie**
This PR competes with https://github.com/logstash-plugins/logstash-input-s3/pull/126 -- by using `java.util.zip.GZipInputStream` alongside `java.io.BufferedReader`, we're able to route around the same problems while maintaining a reasonable memory profile in worst-case, all without requiring explicit configuration.

bonus: `java.util.zip.GZipInputStream` handles multi-compressed streams
transparently, so this simplifies the implementation significantly.

### CODE REVIEW

File: lib/logstash/inputs/s3.rb [hunk 1]
```diff
@@ -253,18 +261,15 @@ def read_plain_file(filename, block)
...
-        io.each_line { |line| block.call(line) }
-        unused = io.unused
-        io.finish
-        break if unused.nil?
-        zio.pos -= unused.length # reset the position to the other block in the stream
-      end
+    fileStream = FileInputStream.new(filename)
+    gzipStream = GZIPInputStream.new(fileStream)
+    decoder = InputStreamReader.new(gzipStream, "UTF-8")
+    buffered = BufferedReader.new(decoder)
```
Comment (jakelandis) [inline lib/logstash/inputs/s3.rb]: Do these Readers/Streams need to be closed ?

**Comment by jakelandis** [comment]
> This PR competes with #126 -- by using java.util.zip.GZipInputStream alongside java.io.BufferedReader, 

+1 to using Java libraries.

File: lib/logstash/inputs/s3.rb [hunk 2]
```diff
@@ -253,18 +261,15 @@ def read_plain_file(filename, block)
...
-        io = Zlib::GzipReader.new(zio)
-        io.each_line { |line| block.call(line) }
-        unused = io.unused
-        io.finish
-        break if unused.nil?
-        zio.pos -= unused.length # reset the position to the other block in the stream
-      end
+    fileStream = FileInputStream.new(filename)
+    gzipStream = GZIPInputStream.new(fileStream)
+    decoder = InputStreamReader.new(gzipStream, "UTF-8")
```
Comment (jakelandis) [inline lib/logstash/inputs/s3.rb]: This presumes text encoding ? Should this be a ByteAraryInputStream ?

Comment (yaauie) [reply lib/logstash/inputs/s3.rb]: The current implementation implicitly assumes UTF-8; we could make this configurable?

**Comment by yaauie** [comment]
@jakelandis good call on closing the `InputReader`s etc; can I get you to take another look?

As for encoding, the current implementation assumes UTF-8, and while we _could_ add a configuration option for which character the file should be read with, I believe doing so is out of scope of this ticket.

**Comment by jakelandis** [comment]
>  I believe doing so is out of scope of this ticket.

agreed. For my own education, does this mean that binary formats + codec are not supported by this plugin ?

Changes as-is LGTM.

Are you going to close #126 ?

**Comment by yaauie** [comment]
> Changes as-is LGTM.

Cool. merge/changelog/version/publish pending.

> Are you going to close #126 ?

I already did.

**Comment by yaauie** [comment]
yeah, that's better:

~~~
                ruby      1.385k (± 6.2%) i/s -      6.930k in   5.022677s
                java     13.892k (± 4.5%) i/s -     69.355k in   5.003254s
~~~

**Pull request merged**